pkg/gator/sync/verify/verify.go

```diff
@@ -0,0 +1,118 @@
+package verify
+
+import (
+	"fmt"
+
+	cfapis "github.com/open-policy-agent/frameworks/constraint/pkg/apis"
+	templatesv1 "github.com/open-policy-agent/frameworks/constraint/pkg/apis/templates/v1"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/core/templates"
+	gkapis "github.com/open-policy-agent/gatekeeper/v3/apis"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager/parser"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/gator/reader"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var scheme *runtime.Scheme
+
+func init() {
+	scheme = runtime.NewScheme()
+	err := cfapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+	err = gkapis.AddToScheme(scheme)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func Verify(unstrucs []*unstructured.Unstructured, flagDiscoveryResults string) (map[string][]int, error) {
+	discoveryResults, err := reader.ReadDiscoveryResults(flagDiscoveryResults)
+	if err != nil {
+		return nil, fmt.Errorf("reading: %w", err)
+	}
+
+	templates := []*templates.ConstraintTemplate{}
+	syncedGVKs := map[schema.GroupVersionKind]struct{}{}
+
+	for _, obj := range unstrucs {
+		if isSyncSet(obj) {
+			syncSet, err := reader.ToSyncSet(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to syncset: %w", obj.GetName(), err)
+			}
+			for _, gvkEntry := range syncSet.Spec.GVKs {
+				gvk := schema.GroupVersionKind{
+					Group:   gvkEntry.Group,
+					Version: gvkEntry.Version,
+					Kind:    gvkEntry.Kind,
+				}
+				if _, exists := discoveryResults[gvk]; exists || discoveryResults == nil {
+					syncedGVKs[gvk] = struct{}{}
+				}
+			}
+		}
+		if isConfig(obj) {
+			config, err := reader.ToConfig(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to config: %w", obj.GetName(), err)
+			}
+			for _, syncOnlyEntry := range config.Spec.Sync.SyncOnly {
+				gvk := schema.GroupVersionKind{
+					Group:   syncOnlyEntry.Group,
+					Version: syncOnlyEntry.Version,
+					Kind:    syncOnlyEntry.Kind,
+				}
+				if _, exists := discoveryResults[gvk]; exists || discoveryResults == nil {
+					syncedGVKs[gvk] = struct{}{}
+				}
+			}
+		}
+		if isTemplate(obj) {
+			templ, err := reader.ToTemplate(scheme, obj)
+			if err != nil {
+				return nil, fmt.Errorf("converting unstructured %q to template: %w", obj.GetName(), err)
+			}
+			templates = append(templates, templ)
+		}
+	}
+
+	missingReqs := map[string][]int{}
+
+	for _, templ := range templates {
+		// Fetch syncrequirements from template
+		syncRequirements, err := parser.ReadSyncRequirements(templ)
+		if err != nil {
+			return nil, fmt.Errorf("reading sync requirements from template %q: %w", templ.GetName(), err)
+		}
```

Comment:
Thinking about the UX here some more -- do we want to err out on the first issue or do we want to best effort look at all templates and see which ones are having issues and which ones are covered, etc.

Reply:
Great idea! Fixed

Reply:
@julianKatz resolving some more merge conflicts between b your review and my edits lol...I ended up continuing on any template errors rather than stopping, so that the eng can see the other templates' issues as well and fix them all at once... I think the only thing that should stop evaluation is config errors, as the templates are independent of each other, wdyt?

Reply:
Personally I like my [proposition](https://github.com/open-policy-agent/gatekeeper/pull/3098#discussion_r1369309334), which differentiates two kinds of errors:

1. "These objects are invalid and I can't interpret them correctly"

2. "These objects are valid but misconfigured"

IMO type 1 should be evaluated first and should stop the command in its tracks.  These errors are kind of "existential", because they prevent correct functioning.  Thus, I think a "fail fast" approach where the error is returned (with no noise about previous templates working or anything like that).

Once everything is known to be valid, errors of type 2 can be collated and printed as a list.  These aren't "errors" as much as they are "valid but negative results".  Thus the differentiation.

Reply:
Synced offline, stop evaluation after assessing validity of all templates if any are invalid